Support arguments with spaces.

If you want to run a task as "Run task with args" and your argument contains spaces within the value, it will not work. Especially if you want to work with the --args from gradle, this is required.

Example:
Your application required an input and output file and can also use "--debug" to output additional debug information. So you want to run the `gradle run` task with the following arguments:
`--args="input.txt -o output.txt --debug"`

Currently you would get the following output:

```text
> Executing task: gradle: run --args="input.txt -o output.txt --debug" <

<-------------> 0% WAITING

FAILURE: Build failed with an exception.

* What went wrong:
Problem configuring task :run from command line.
> Unknown command-line option '-o'.

* Try:
Run gradle help --task :run to get task usage details. Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Could not execute build using Gradle distribution 'https://services.gradle.org/distributions/gradle-6.5.1-bin.zip'.
```

--- a/extension/package.json
+++ b/extension/package.json
@@ -751,6 +751,7 @@
     "rollup-plugin-terser": "^6.1.0",
     "sinon": "^9.0.2",
     "snyk": "^1.366.2",
+    "string-argv": "^0.3.1",
     "ts-protoc-gen": "^0.12.0",
     "typescript": "^3.9.7",
     "vsce": "^1.77.0",

--- a/extension/src/commands/runBuildCommand.ts
+++ b/extension/src/commands/runBuildCommand.ts
@@ -1,4 +1,5 @@
 import * as vscode from 'vscode';
+import { parseArgsStringToArgv } from 'string-argv';
 import { getGradleCommand, getRootProjectFolder } from '../input';
 import { GradleRunnerTerminal } from '../terminal';
 import { getRunBuildCancellationKey } from '../client/CancellationKeys';
@@ -13,7 +14,8 @@ export async function runBuildCommand(): Promise<void> {
   if (!gradleCommand) {
     return;
   }
-  const args: string[] = gradleCommand.split(' ').filter(Boolean);
+
+  const args: string[] = parseArgsStringToArgv(gradleCommand.trim());
   const cancellationKey = getRunBuildCancellationKey(
     rootProject.getProjectUri().fsPath,
     args

--- a/extension/src/tasks/taskUtil.ts
+++ b/extension/src/tasks/taskUtil.ts
@@ -1,4 +1,5 @@
 import * as vscode from 'vscode';
+import { parseArgsStringToArgv } from 'string-argv';
 import { GradleTask, GradleProject, GradleBuild } from '../proto/gradle_pb';
 import { TaskArgs } from '../stores/types';
 import { Extension } from '../extension';
@@ -150,9 +151,9 @@ export function createTaskFromDefinition(
   rootProject: RootProject
 ): vscode.Task {
   const taskTerminalsStore = Extension.getInstance().getTaskTerminalsStore();
-  const args = [definition.script].concat(
-    definition.args.split(' ').filter(Boolean)
-  );
+  const args = [definition.script]
+    .concat(parseArgsStringToArgv(definition.args.trim()))
+    .filter(Boolean);
   const taskName = buildTaskName(definition);
   const cancellationKey = getRunTaskCommandCancellationKey(
     rootProject.getProjectUri().fsPath,

--- a/extension/src/test/integration/gradle/extension.test.ts
+++ b/extension/src/test/integration/gradle/extension.test.ts
@@ -88,10 +88,10 @@ describe(getSuiteName('Extension'), () => {
       );
     });
 
-    it('should run a gradle task with custom args', async () => {
+    it('should run a gradle task with custom args that contain spaces', async () => {
       sinon
         .stub(vscode.window, 'showInputBox')
-        .returns(Promise.resolve('-PcustomProp=foo'));
+        .returns(Promise.resolve('-PcustomProp="foo bar"'));
 
       assert.ok(extension);
 
@@ -120,7 +120,9 @@ describe(getSuiteName('Extension'), () => {
           treeItem
         );
       });
-      assert.ok(spy.calledWith(sinon.match('Hello, Project Property!foo')));
+      assert.ok(
+        spy.calledWith(sinon.match('Hello, Project Property!"foo bar"'))
+      );
     });
   });
 

